File: test/_harnesses/csharp/sized.cs
using System;
using System.IO;
using System.Collections.Generic;

using SizedMessage;

class SizedHarness: TestHarness {

    static void Main(string[] args) {
        Dictionary<string, string> parsedArgs = parseArguments(args);

        var shortList = new SizedMessage.TextContainer();
        shortList.label = "list that fits in a byte";
        shortList.collection = new List<string> {
            "Lorem", "ipsum", "dolor", "sit", "amet", "consectetur",
            "adipiscing", "elit", "sed", "do", "eiusmod", "tempor",
            "incididunt", "ut", "labore", "et", "dolore", "magna",
            "aliqua", "Ut", "enim", "ad", "minim", "veniam",
            "quis", "nostrud", "exercitation", "ullamco", "laboris", "nisi",
            "ut", "aliquip", "ex", "ea", "commodo", "consequat",
            "Duis", "aute", "irure", "dolor", "in", "reprehenderit",
            "in", "voluptate", "velit", "esse", "cillum", "dolore",
            "eu", "fugiat", "nulla", "pariatur", "Excepteur", "sint",
            "occaecat", "cupidatat", "non", "proident", "sunt", "in",
            "culpa", "qui", "officia", "deserunt", "mollit", "anim",
            "id", "est", "laborum",
        };

        if (parsedArgs.ContainsKey("generate"))
        {
            string outPath = parsedArgs["generate"];
            string outDir = System.IO.Path.GetDirectoryName(outPath);
            System.IO.Directory.CreateDirectory(outDir);
            FileStream f = new FileStream(outPath, FileMode.Create);
            BinaryWriter bw = new BinaryWriter(f);
            shortList.WriteBytes(bw, false);

            softAssert(shortList.GetSizeInBytes() == 464, "short list size calculation check");
            softAssert(shortList.GetSizeInBytes() == bw.BaseStream.Position, "written bytes check");
        }
        else if (parsedArgs.ContainsKey("read"))
        {
            FileStream f = File.OpenRead(parsedArgs["read"]);
            BinaryReader br = new BinaryReader(f);

            var input = SizedMessage.TextContainer.FromBytes(br);

            softAssert(input.label == shortList.label, "readback label comparison");
            softAssert(input.collection.Count == shortList.collection.Count, "readback list length");
            for (int i = 0; i < input.collection.Count; i++)
            {
                softAssert(input.collection[i] == shortList.collection[i], "short list comparison");
            }
        }


        check();
    }
}

File: test/_harnesses/rust/src/sized.rs
// message files will include code that's not necessarily called in each test
#![allow(dead_code)]

use std::fs;

// in a real program you would want this file to have an appropriate name
//   but for testing I'd rather not deal with it having a different name
//   than all the other files.
#[allow(non_snake_case)]
mod SizedMessage;
use SizedMessage::*;

mod util;

fn main() {
    let mut checker = util::Checker { ok: true };
    let args = util::arg_parse();

    let short_list = TextContainer {
        label: "list that fits in a byte".to_string(),
        collection: vec![
            "Lorem".to_string(), "ipsum".to_string(), "dolor".to_string(), "sit".to_string(), "amet".to_string(), "consectetur".to_string(),
            "adipiscing".to_string(), "elit".to_string(), "sed".to_string(), "do".to_string(), "eiusmod".to_string(), "tempor".to_string(),
            "incididunt".to_string(), "ut".to_string(), "labore".to_string(), "et".to_string(), "dolore".to_string(), "magna".to_string(),
            "aliqua".to_string(), "Ut".to_string(), "enim".to_string(), "ad".to_string(), "minim".to_string(), "veniam".to_string(),
            "quis".to_string(), "nostrud".to_string(), "exercitation".to_string(), "ullamco".to_string(), "laboris".to_string(), "nisi".to_string(),
            "ut".to_string(), "aliquip".to_string(), "ex".to_string(), "ea".to_string(), "commodo".to_string(), "consequat".to_string(),
            "Duis".to_string(), "aute".to_string(), "irure".to_string(), "dolor".to_string(), "in".to_string(), "reprehenderit".to_string(),
            "in".to_string(), "voluptate".to_string(), "velit".to_string(), "esse".to_string(), "cillum".to_string(), "dolore".to_string(),
            "eu".to_string(), "fugiat".to_string(), "nulla".to_string(), "pariatur".to_string(), "Excepteur".to_string(), "sint".to_string(),
            "occaecat".to_string(), "cupidatat".to_string(), "non".to_string(), "proident".to_string(), "sunt".to_string(), "in".to_string(),
            "culpa".to_string(), "qui".to_string(), "officia".to_string(), "deserunt".to_string(), "mollit".to_string(), "anim".to_string(),
            "id".to_string(), "est".to_string(), "laborum".to_string(),
        ]
    };

    if args.contains_key("generate") {
        let mut writer: Vec<u8> = Vec::new();
        short_list.write_bytes(&mut writer, false);

        let filename = args.get("generate").unwrap();
        checker.soft_assert(short_list.get_size_in_bytes() == 464, "short list size calculation check");
        checker.soft_assert(writer.len() as u32 == short_list.get_size_in_bytes(), "written bytes check");
        fs::write(filename, writer).unwrap();
    }
    else if args.contains_key("read") {
        let filename = args.get("read").unwrap();
        let buffer = fs::read(&filename).unwrap();
        let mut reader = BufferReader::new(buffer);
        let input = TextContainer::from_bytes(&mut reader).unwrap();


        checker.soft_assert(input.label == short_list.label, "readback label comparison");
        checker.soft_assert(input.collection.len() == short_list.collection.len(), "readback list length");
        for i in 0..input.collection.len() {
            checker.soft_assert(input.collection[i] == short_list.collection[i], "short list comparison");
        }

    }

    checker.check();
}
